amivapi/tests/auth/test_oauth.py

```diff
@@ -0,0 +1,154 @@
+            '/oauth/authorize?'
+            'response_type=token'
+            '&client_id=TestClient'
+            '&redirect_uri=https://bla.org/tool?a=b'
+            '&state=xyz', status_code=200)
+
+        # As we have a valid token already it should not ask for a username
+        # and password
+        assert b'Username: ' not in login_page.data
+
+        # Simulate sending the login form
+        response = self.api.post(
+            '/oauth/login', data={
+                'response_type': 'token',
+                'client_id': 'TestClient',
+                'redirect_uri': 'https://bla.org/tool?a=b',
+                'state': 'xyz'
+            }, headers={'content-type': 'application/x-www-form-urlencoded'},
+            status_code=302)
+
+        parsed_url = urlsplit(response.location)
+        assert parsed_url.scheme == 'https'
+        assert parsed_url.netloc == 'bla.org'
+        assert parsed_url.path == '/tool'
+
+        query_params = dict(parse_qsl(parsed_url.query))
+        assert query_params['a'] == 'b'
+        assert query_params['access_token'] == token
+        assert query_params['token_type'] == 'bearer'
+        assert query_params['state'] == 'xyz'
+
+    def test_login(self):
+        """Test that a user can login with username and password."""
+        user = self.new_object('users', nethz='testuser', password='pass')
+        token = self.get_user_token(user['_id'])
+
+        login_page = self.api.get(
+            '/oauth/authorize?'
+            'response_type=token'
+            '&client_id=TestClient'
+            '&redirect_uri=https://bla.org/tool?a=b'
+            '&state=xyz', status_code=200)
+
+        # As we have a valid token already it should not ask for a username
+        # and password
+        assert b'Username: ' in login_page.data
+
+        # Simulate sending the login form
+        response = self.api.post(
+            '/oauth/login', data={
+                'response_type': 'token',
+                'client_id': 'TestClient',
+                'redirect_uri': 'https://bla.org/tool?a=b',
+                'state': 'xyz',
+                'username': 'testuser',
+                'password': 'pass'
+            }, headers={'content-type': 'application/x-www-form-urlencoded'},
+            status_code=302)
+
+        parsed_url = urlsplit(response.location)
+        assert parsed_url.scheme == 'https'
+        assert parsed_url.netloc == 'bla.org'
+        assert parsed_url.path == '/tool'
+
+        query_params = dict(parse_qsl(parsed_url.query))
+        assert query_params['a'] == 'b'
+        assert query_params['token_type'] == 'bearer'
+        assert query_params['state'] == 'xyz'
+
+        # Check the token works
+        token = query_params['access_token']
+        resp = self.api.get('/sessions?where={"token":"%s"}' % token,
+                            token=token, status_code=200).json
+        assert len(resp['_items']) == 1
+
+    def test_login_wrong_data(self):
```

Comment:
should add the case of wrong username or empty fields

Reply:
done